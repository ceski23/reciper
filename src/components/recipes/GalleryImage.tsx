/* eslint-disable @typescript-eslint/no-use-before-define */
import styled from '@emotion/styled';
import { FC } from 'react';
import { Item } from 'react-photoswipe-gallery';

import { useImageDimensions } from 'hooks/useImageDimensions';

interface GalleryImageProps {
  url: string;
}

export const GalleryImage: FC<GalleryImageProps> = ({ url }) => {
  const dimensions = useImageDimensions(url);

  return (
    <Item
      original={url}
      thumbnail={url}
      key={url}
      width={dimensions?.width}
      height={dimensions?.height}
      cropped
    >
      {({ ref, open }) => (
        <ImageThumbnail
          ref={ref as React.MutableRefObject<HTMLImageElement>}
          onClick={open}
          src={url}
        />
      )}
    </Item>
  );
};

const ImageThumbnail = styled.img`
  width: 100%;
  height: auto;
  aspect-ratio: 1;
  object-fit: cover;
  border-radius: 10px;
  cursor: pointer;
`;
