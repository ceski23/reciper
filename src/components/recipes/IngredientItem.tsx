import { css, Theme } from '@emotion/react';
import styled from '@emotion/styled';
import { VFC } from 'react';

import { ReactComponent as GroceryIcon } from 'assets/recipes/grocery.svg';

import { useAppSelector } from 'hooks/store';

import ingredientsDatabase from 'services/ingredients/database';
import { ParsedIngredient } from 'services/ingredients/models';
import { units } from 'services/units';
import { convertIngredient } from 'services/units/utils';

import { selectConversionPrecision, selectUnitsConversions, selectUseUnitsConversion } from 'store/settings';

import { color } from 'utils/styles/theme';
import { fluidTypography } from 'utils/styles/typography';

interface Props {
  ingredient: ParsedIngredient;
  onClick?: () => void;
}

const iconStyles = ({ theme }: { theme: Theme }) => css`
  border-radius: 10px;
  background-color: ${theme.colors.backgroundAlt};
  width: 50px;
  height: 50px;
  border: none;
  margin-right: 20px;
  flex-shrink: 0;
  padding: 10px;
`;

const Icon = styled.img`
  ${iconStyles}

  object-fit: contain;
`;

const FallbackIcon = styled(GroceryIcon)`
  ${iconStyles}

  fill: ${color('text')};
`;

const Wrapper = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
`;

const Text = styled.p`
  flex-grow: 1;
  margin: 0;
  ${({ theme }) => fluidTypography(
    theme.breakpoints.small,
    theme.breakpoints.xlarge,
    16,
    20,
  )}
`;

const formatQuantity = (quantity: number, precision?: number) => {
  if (precision === undefined) return quantity;
  return Number(quantity.toFixed(precision));
};

export const IngredientItem: VFC<Props> = ({ ingredient, onClick }) => {
  const conversions = useAppSelector(selectUnitsConversions);
  const conversionEnabled = useAppSelector(selectUseUnitsConversion);
  const convertionPrecision = useAppSelector(selectConversionPrecision);

  const precision = conversionEnabled ? convertionPrecision : undefined;

  const convertedIngredient = conversionEnabled ? convertIngredient(
    ingredient,
    ('type' in ingredient && ingredient.type) ? conversions[ingredient.type] : undefined,
  ) : ingredient;

  const quantity = ('parsed' in convertedIngredient)
    ? formatQuantity(convertedIngredient.quantity, precision)
    : undefined;

  const unitText = ('unit' in convertedIngredient)
    ? units[convertedIngredient.unit].formatter.format({
      quantity: formatQuantity(convertedIngredient.quantity, precision),
    })
    : undefined;

  return (
    <Wrapper onClick={onClick}>
      {('type' in ingredient && ingredient.type) ? (
        <Icon src={ingredientsDatabase[ingredient.type].image} />
      ) : (
        <FallbackIcon />
      )}

      <Text>
        {('parsed' in ingredient) ? (
          <>
            {ingredient.parsed.begin}{' '}
            <strong>{quantity} {unitText}</strong>
            {' '}{ingredient.parsed.end}
          </>
        ) : ingredient.original}
      </Text>
    </Wrapper>
  );
};
