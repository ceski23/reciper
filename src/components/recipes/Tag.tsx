import styled from '@emotion/styled';
import { reverse } from 'named-urls';
import { VFC } from 'react';

import { Styleable } from 'types';

import { Link } from 'components/common/Link';

import { urls } from 'routing/urls';

import { color } from 'utils/styles/theme';
import { fluidTypography } from 'utils/styles/typography';

const Container = styled(Link)`
  background-color: ${color('backgroundHover')};
  border-radius: 10px;
  padding: 10px;
  color: ${color('primary')};
  ${({ theme }) => fluidTypography(
    theme.breakpoints.small,
    theme.breakpoints.xlarge,
    14,
    16,
  )}
`;

interface Props {
  tag: string;
}

export const Tag: VFC<Props & Styleable> = ({ tag, className }) => (
  <Container
    className={className}
    to={reverse(urls.tags.tag, { tag: encodeURIComponent(tag) })}
  >
    {tag}
  </Container>
);
