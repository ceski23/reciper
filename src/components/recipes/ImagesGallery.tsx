/* eslint-disable @typescript-eslint/no-use-before-define */
import styled from '@emotion/styled';
import { FC } from 'react';
import { Gallery } from 'react-photoswipe-gallery';

import { GalleryImage } from 'components/recipes/GalleryImage';

interface ImagesGalleryProps {
  images: string[]
}

export const ImagesGallery: FC<ImagesGalleryProps> = ({ images }) => (
  <div>
    <h2>Galeria</h2>
    <Gallery>
      <ImagesContainer>
        {images.map((image) => (
          <GalleryImage key={image} url={image} />
        ))}
      </ImagesContainer>
    </Gallery>
  </div>
);

export default ImagesGallery;

const ImagesContainer = styled.div`
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(100px, 1fr));
  gap: 20px;
`;
