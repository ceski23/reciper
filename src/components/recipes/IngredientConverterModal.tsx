import styled from '@emotion/styled';
import { ChangeEvent, FC } from 'react';

import { ReactComponent as ConversionIcon } from 'assets/recipes/conversion.svg';

import { Modal, ModalProps } from 'components/common/modal/Modal';
import { Field } from 'components/forms/inputs/Field';

import { useIngredientConversion } from 'hooks/recipes/useIngredientConversion';

import KNOWN_INGREDIENTS from 'services/ingredients/database';
import { IngredientWithUnitAndType } from 'services/ingredients/models';
import { UnitName, units } from 'services/units';

import { color } from 'utils/styles/theme';

const IngredientRow = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: 20px;
  font-weight: 700;
  margin-bottom: 20px;
`;

const IngredientImage = styled.img`
  width: 50px;
  height: 50px;
  object-fit: contain;
`;

const StyledConversionIcon = styled(ConversionIcon)`
  width: 50px;
  height: 50px;
  fill: ${color('textalt')};
  flex-grow: 1;
  flex-shrink: 0;
`;

const ConversionRow = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: 10px;
`;

const Select = styled.select`
  border: none;
  background: none;
  font-size: inherit;
  color: inherit;
`;

const ConversionContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  row-gap: 10px;
`;

const QuantityField = styled(Field)`
  max-width: 100px;
  min-width: 50px;
  text-align: end;

  &::-webkit-outer-spin-button,
  &::-webkit-inner-spin-button {
      /* display: none; <- Crashes Chrome on hover */
      -webkit-appearance: none;
      margin: 0; /* <-- Apparently some margin are still there even though it's hidden */
  }

  &[type=number] {
      -moz-appearance:textfield; /* Firefox */
  }
`;

const capitalize = (text: string) => (
  `${text[0].toLocaleUpperCase()}${text.substring(1)}`
);

interface IngredientConverterModalProps {
  ingredient: IngredientWithUnitAndType;
}

const stringToNumber = (text: string) => {
  const num = Number.parseFloat(text);
  return Number.isNaN(num) ? 0 : num;
};

export const IngredientConverterModal: FC<IngredientConverterModalProps & ModalProps> = ({
  ingredient, ...props
}) => {
  const availableUnits = Object.keys(KNOWN_INGREDIENTS[ingredient.type].conversions) as UnitName[];
  const ingredientObject = KNOWN_INGREDIENTS[ingredient.type];
  const {
    changeQuantity, changeUnit, conversionData, swapPositions,
  } = useIngredientConversion(ingredient);

  const handleQuantityChange = (event: ChangeEvent<HTMLInputElement>) => {
    const fieldName = event.currentTarget.name;
    const quantity = stringToNumber(event.currentTarget.value);

    if (fieldName === 'from') changeQuantity({ from: quantity });
    else if (fieldName === 'to') changeQuantity({ to: quantity });
  };

  const handleUnitChange = (event: ChangeEvent<HTMLSelectElement>) => {
    const fieldName = event.currentTarget.name;
    const unit = event.currentTarget.value as UnitName;

    if (fieldName === 'from') changeUnit({ from: unit });
    else if (fieldName === 'to') changeUnit({ to: unit });
  };

  return (
    <Modal {...props}>
      <Modal.Header title="Kalkulator jednostek" />
      <Modal.Body>
        <IngredientRow>
          <IngredientImage src={ingredientObject.image} alt={ingredientObject.name} />
          <p>{capitalize(ingredientObject.name)}</p>
        </IngredientRow>

        <ConversionContainer>
          <ConversionRow>
            <QuantityField
              id="fromQuantity"
              type="number"
              name="from"
              inputMode="decimal"
              value={conversionData.quantity.from}
              onChange={handleQuantityChange}
              min={0}
            />

            <Select value={conversionData.unit.from} onChange={handleUnitChange} name="from">
              {availableUnits.map((unit) => (
                <option key={unit} value={unit}>
                  {units[unit].formatter.format({ quantity: conversionData.quantity.from })}
                </option>
              ))}
            </Select>
          </ConversionRow>

          <StyledConversionIcon onClick={swapPositions} />

          <ConversionRow>
            <QuantityField
              id="fromQuantity"
              type="number"
              name="to"
              inputMode="decimal"
              value={conversionData.quantity.to}
              onChange={handleQuantityChange}
              min={0}
            />

            <Select value={conversionData.unit.to} onChange={handleUnitChange} name="to">
              {availableUnits.map((unit) => (
                <option key={unit} value={unit}>
                  {units[unit].formatter.format({ quantity: conversionData.quantity.to })}
                </option>
              ))}
            </Select>
          </ConversionRow>
        </ConversionContainer>
      </Modal.Body>
      <Modal.Footer cancelText="Zamknij" />
    </Modal>
  );
};
