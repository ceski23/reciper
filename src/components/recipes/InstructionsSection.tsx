/* eslint-disable @typescript-eslint/no-use-before-define */
import styled from '@emotion/styled';
import { FC, useState } from 'react';

import { RecipeStep } from 'components/recipes/RecipeStep';

import { RecipeInstruction } from 'services/recipes';

export interface InstructionsGroup {
  group?: string;
  instructions: RecipeInstruction[];
}

interface InstructionsSectionProps {
  instructionsGroup: InstructionsGroup;
}

export const InstructionsSection: FC<InstructionsSectionProps> = ({
  instructionsGroup,
}) => {
  const [doneSteps, setDoneSteps] = useState<number[]>([]);

  const handleStepClicked = (idx: number) => {
    if (doneSteps.includes(idx)) {
      setDoneSteps(doneSteps.filter((i) => i !== idx));
    } else {
      setDoneSteps([...doneSteps, idx]);
    }
  };

  return (
    <>
      <h2>{instructionsGroup.group || 'Instrukcje'}</h2>

      <RecipeSteps>
        {instructionsGroup.instructions.map((ins, idx) => (
          <RecipeStep
          // eslint-disable-next-line react/no-array-index-key
            key={idx}
            stepNumber={idx + 1}
            instruction={ins.text}
            done={doneSteps.includes(idx)}
            onClick={() => handleStepClicked(idx)}
          />
        ))}
      </RecipeSteps>
    </>
  );
};

const RecipeSteps = styled.div`
  display: flex;
  flex-direction: column;
  gap: 20px;

  &:not(:last-child) {
    margin-bottom: 40px;
  }
`;
