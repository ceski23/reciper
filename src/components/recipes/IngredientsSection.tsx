/* eslint-disable @typescript-eslint/no-use-before-define */
import styled from '@emotion/styled';
import IntlMessageFormat from 'intl-messageformat';
import { FC, useMemo, useState } from 'react';
import toast from 'react-hot-toast';

import { ReactComponent as ShoppingList } from 'assets/recipes/shopping-list.svg';

import { Button } from 'components/common/Button';
import { IngredientItem } from 'components/recipes/IngredientItem';
import { Servings } from 'components/recipes/Servings';

import { useAccountProvider } from 'hooks/accounts/useAccountProvider';
import { useAppSelector } from 'hooks/store';

import { ParsedIngredient } from 'services/ingredients/models';
import { parseIngredient } from 'services/ingredients/parser';
import { RecipeIngredient } from 'services/recipes';

import { selectShoppingList } from 'store/user';

import { color } from 'utils/styles/theme';
import { fluidTypography } from 'utils/styles/typography';

interface IngredientsGroup {
  group?: string;
  ingredients: RecipeIngredient[];
}

interface IngredientsSectionProps {
  ingredientsGroup: IngredientsGroup;
  servings?: number;
  onIngredientClick?: (ingredient: ParsedIngredient) => void;
  recipeName: string;
}

export const ingredientsText = new IntlMessageFormat(`
  {quantity, plural,
    one {# produkt}
    few {# produkty}
    many {# produktów}
    other {# produktu}
  }
`, 'pl-PL');

const adjustIngredientQuantity = (
  ingredient: ParsedIngredient,
  servings?: number,
  initialServings?: number,
) => {
  if (!servings || !('quantity' in ingredient) || !initialServings) return ingredient;
  return {
    ...ingredient,
    quantity: (ingredient.quantity / initialServings) * servings,
  };
};

export const IngredientsSection: FC<IngredientsSectionProps> = ({
  ingredientsGroup, servings: initialServings, onIngredientClick, recipeName,
}) => {
  const [servings, setServings] = useState(initialServings);
  const shoppingList = useAppSelector(selectShoppingList);
  const accountProvider = useAccountProvider();

  const parsedIngredients = useMemo<ParsedIngredient[]>(() => (
    ingredientsGroup.ingredients
      .map((ingredient) => parseIngredient(ingredient.text))
      .map((ingredient) => adjustIngredientQuantity(ingredient, servings, initialServings))
  ), [ingredientsGroup.ingredients, initialServings, servings]);

  const handleAddToShoppingList = () => {
    if (accountProvider && shoppingList && parsedIngredients) {
      const promise = accountProvider.addIngredientsToList(
        shoppingList.id,
        parsedIngredients.map((i) => i.original),
        ingredientsGroup.group ? `${recipeName} - ${ingredientsGroup.group}` : recipeName,
      );

      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      toast.promise(promise, {
        loading: 'Dodawanie do listy',
        success: 'Dodano składniki do listy',
        error: 'Wystąpił błąd podczas dodawania składników',
      });
    }
  };

  return (
    <>
      <IngredientsHeader>
        <h2>{ingredientsGroup.group || 'Składniki'}</h2>

        <AltText>
          {ingredientsText.format({ quantity: ingredientsGroup.ingredients.length })}
        </AltText>
      </IngredientsHeader>

      <IngredientsContainer>
        {!!servings && <Servings servings={servings} onServingsChange={setServings} />}

        <IngredientsList>
          {parsedIngredients.map((ingredient) => (
            <IngredientItem
              ingredient={ingredient}
              key={ingredient.original}
              onClick={() => onIngredientClick?.(ingredient)}
            />
          ))}
        </IngredientsList>

        {shoppingList && (
        <AddToListButton icon={ShoppingList} onClick={handleAddToShoppingList}>
          {`Dodaj do "${shoppingList.name}"`}
        </AddToListButton>
        )}
      </IngredientsContainer>
    </>
  );
};

const IngredientsHeader = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

const AltText = styled.span`
  color: ${color('textalt')};
  ${({ theme }) => fluidTypography(
    theme.breakpoints.small,
    theme.breakpoints.xlarge,
    14,
    16,
  )}
`;

const IngredientsContainer = styled.div`
  display: flex;
  flex-direction: column;
  gap: 20px;

  &:not(:last-child) {
    margin-bottom: 20px;
  }
`;

const IngredientsList = styled.div`
  display: flex;
  flex-direction: column;
  gap: 10px;
`;

const AddToListButton = styled(Button)`
  width: 100%;
`;
