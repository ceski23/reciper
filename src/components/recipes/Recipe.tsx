import { ThemeProvider, useTheme } from '@emotion/react/macro';
import styled from '@emotion/styled/macro';
import dayjs from 'dayjs';
import IntlMessageFormat from 'intl-messageformat';
import { reverse } from 'named-urls';
import {
  VFC, useMemo, MouseEventHandler, useState,
} from 'react';
import toast from 'react-hot-toast';
import { useNavigate } from 'react-router';

import { ReactComponent as DisketteIcon } from 'assets/common/diskette.svg';
import { ReactComponent as PencilIcon } from 'assets/common/pencil.svg';
import { ReactComponent as TrashIcon } from 'assets/common/trash.svg';
import { ReactComponent as ClockIcon } from 'assets/recipes/clock.svg';
import { ReactComponent as StarIcon } from 'assets/recipes/favourite.svg';
import { ReactComponent as FlameIcon } from 'assets/recipes/fire.svg';
import { ReactComponent as ServingsIcon } from 'assets/recipes/servings.svg';
import { ReactComponent as ShoppingList } from 'assets/recipes/shopping-list.svg';

import { Button } from 'components/common/Button';
import { FluidContainer } from 'components/common/Container';
import { IngredientItem } from 'components/recipes/IngredientItem';
import { RecipeCover } from 'components/recipes/RecipeCover';
import { RecipeFeature } from 'components/recipes/RecipeFeature';
import { RecipeProvider } from 'components/recipes/RecipeProvider';
import { RecipeStep } from 'components/recipes/RecipeStep';
import { Servings, servingsText } from 'components/recipes/Servings';
import { Tag } from 'components/recipes/Tag';

import { useAccountProvider } from 'hooks/accounts/useAccountProvider';
import { useAppDispatch, useAppSelector } from 'hooks/store';

import { urls } from 'routing/urls';

import { ParsedIngredient, parseIngredient } from 'services/ingredients/parser';
import { Recipe as RecipeType } from 'services/recipes';
import { chooseProvider } from 'services/recipes/providers';

import {
  removeRecipeById, removeRecipeByUrl, saveRecipe, selectRecipes,
} from 'store/recipes';
import { selectDynamicPrimaryColor } from 'store/settings';
import { selectShoppingList } from 'store/user';

import { media } from 'utils/styles/mediaQueries';
import { lightTheme, generateThemeColors } from 'utils/styles/theme';
import { fluidTypography } from 'utils/styles/typography';
import { normalizeUrl } from 'utils/url';

export const ingredientsText = new IntlMessageFormat(`
  {quantity, plural,
    one {# produkt}
    few {# produkty}
    many {# produktów}
    other {# produktu}
  }
`, 'pl-PL');

interface Props {
  recipe: RecipeType
}

const Title = styled('h1')`
  margin-bottom: 10px;
  ${({ theme }) => fluidTypography(
    theme.breakpoints.small,
    theme.breakpoints.xlarge,
    25,
    45,
  )}
`;

const RecipeSteps = styled.div`
  display: flex;
  flex-direction: column;
  gap: 20px;
`;

const Features = styled.div`
  display: grid;
  justify-content: space-around;
  grid-auto-flow: column;
  grid-template-columns: repeat(3, 1fr);
`;

const Description = styled.div`
  font-size: clamp(16px, 1.1vw, 26px);
  line-height: 1.4;
  text-align: justify;
`;

const ContentWrapper = styled(FluidContainer)`
  margin: 50px 0;
  color: ${(props) => props.theme.colors.text};
  background-color: ${(props) => props.theme.colors.background};
  row-gap: 50px;

  ${media.down('medium')} {
    margin-top: 0;
  }
`;

const AddToListButton = styled(Button)`
  width: 100%;
`;

const SaveIcon = styled(DisketteIcon)`
  width: 20px;
  height: 20px;
`;

const UnsaveIcon = styled(TrashIcon)`
  width: 20px;
  height: 20px;
`;

const IngredientsHeader = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

const AltText = styled.span`
  color: ${(props) => props.theme.colors.textalt};
  ${({ theme }) => fluidTypography(
    theme.breakpoints.small,
    theme.breakpoints.xlarge,
    14,
    16,
  )}
`;

const RecipeLink = styled.a`
  text-decoration: none;
  color: unset;
`;

const Tags = styled.div`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  gap: 10px;

  & > * {
    text-align: center;
    flex: 1 0 auto;
  }
`;

const IngredientsContainer = styled.div`
  display: flex;
  flex-direction: column;
  gap: 10px;
`;

const IngredientsSection = styled.div`
  display: flex;
  flex-direction: column;
  position: sticky;
  top: 20px;
  margin-bottom: auto;
  gap: 30px;

  ${media.down('large')} {
    position: static;
  }
`;

const SideBySide = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr;
  gap: 100px;

  ${media.down('large')} {
    display: flex;
    flex-direction: column-reverse;
    gap: 40px;
  }
`;

const Actions = styled.div`
  display: flex;
  flex-direction: row;
  gap: 30px;

  & > * {
    flex: 1;
  }

  ${media.down('small')} {
    flex-direction: column-reverse;
  }
`;

export const Recipe: VFC<Props> = ({ recipe }) => {
  const recipeProvider = recipe.url ? chooseProvider(recipe.url) : undefined;
  const dispatch = useAppDispatch();
  const [doneSteps, setDoneSteps] = useState<number[]>([]);
  const theme = useTheme();
  const shoppingList = useAppSelector(selectShoppingList);
  const accountProvider = useAccountProvider();
  const dynamicColor = useAppSelector(selectDynamicPrimaryColor);
  const [servings, setServings] = useState(recipe.servings);
  const recipes = useAppSelector(selectRecipes);
  const navigate = useNavigate();

  const isRecipeSaved = useMemo(() => {
    if (recipe.url === undefined) return true;
    const recipeUrl = recipe.url;

    return Object
      .values(recipes)
      .some((r) => r.url === normalizeUrl(recipeUrl));
  }, [recipe.url, recipes]);

  const parsedIngredients = useMemo<ParsedIngredient[]>(() => (
    (recipe?.ingredients ?? [])
      .map(parseIngredient)
      .map((ingredient) => {
        if (!servings || !('quantity' in ingredient) || !recipe.servings) return ingredient;
        return { ...ingredient, quantity: (ingredient.quantity / recipe.servings) * servings };
      })
  ), [recipe?.ingredients, recipe.servings, servings]);

  const handleStepClicked = (idx: number) => {
    if (doneSteps.includes(idx)) setDoneSteps(doneSteps.filter((i) => i !== idx));
    else setDoneSteps([...doneSteps, idx]);
  };

  const preparationDuration = useMemo(() => {
    if (!recipe.prepTime) return undefined;
    return dayjs.duration(recipe.prepTime).locale('pl').humanize();
  }, [recipe.prepTime]);

  const colorizedTheme = useMemo(() => {
    if (!recipe.color || !dynamicColor) return undefined;
    return { ...lightTheme, colors: generateThemeColors(recipe.color, theme.type) };
  }, [dynamicColor, recipe.color, theme.type]);

  const handleSaveRecipe: MouseEventHandler<HTMLButtonElement> = () => {
    dispatch(saveRecipe(recipe));
    navigate(
      reverse(urls.recipes.recipeById, { recipeId: recipe.id }),
      { replace: true },
    );
    toast.success('Przepis został zapisany');
  };

  const handleRemoveRecipe: MouseEventHandler<HTMLButtonElement> = () => {
    if (recipe.url) dispatch(removeRecipeByUrl(recipe.url));
    else dispatch(removeRecipeById(recipe.id));

    if (recipe.url) {
      navigate(
        reverse(
          urls.recipes.recipeByUrl,
          { recipeUrl: encodeURIComponent(recipe.url) },
        ),
      );
    } else navigate(urls.home);

    toast.success('Przepis został usunięty');
  };

  const handleEditRecipe = () => {
    navigate(reverse(urls.recipes.edit, { recipeId: recipe.id }));
  };

  const handleAddToShoppingList = () => {
    if (accountProvider && shoppingList && parsedIngredients) {
      const promise = accountProvider.addIngredientsToList(
        shoppingList.id,
        parsedIngredients.map((i) => i.original),
        recipe.name,
      );

      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      toast.promise(promise, {
        loading: 'Dodawanie do listy',
        success: 'Dodano składniki do listy',
        error: 'Wystąpił błąd podczas dodawania składników',
      });
    }
  };

  return (
    <ThemeProvider theme={colorizedTheme ?? theme}>
      <ContentWrapper>
        {recipe.image && <RecipeCover src={recipe.image} />}

        <div>
          <Title>{recipe.name}</Title>
          {recipe.url && recipeProvider && (
            <RecipeProvider provider={recipeProvider} recipeUrl={recipe.url} />
          )}
        </div>

        {(recipe.rating || recipe.calories || preparationDuration) && (
          <Features>
            {preparationDuration && <RecipeFeature icon={ClockIcon} text={preparationDuration} />}
            {recipe.calories && <RecipeFeature icon={FlameIcon} text={`${recipe.calories} kcal`} />}
            {recipe.rating && <RecipeFeature icon={StarIcon} text={recipe.rating.toFixed(1)} />}
            {recipe.servings && (
              <RecipeFeature
                icon={ServingsIcon}
                text={servingsText.format({ quantity: recipe.servings }).toString()}
              />
            )}
          </Features>
        )}

        {recipe.description && (
          <Description>{recipe.description}</Description>
        )}

        {recipe.tags.length > 0 && (
          <Tags>{recipe.tags.map((tag) => <Tag key={tag} tag={tag} />)}</Tags>
        )}

        {isRecipeSaved ? (
          <Actions>
            <Button icon={UnsaveIcon} onClick={handleRemoveRecipe}>Usuń przepis</Button>
            <Button icon={PencilIcon} onClick={handleEditRecipe}>Edytuj przepis</Button>
          </Actions>
        ) : (
          <Button icon={SaveIcon} onClick={handleSaveRecipe}>Zapisz przepis</Button>
        )}

        <SideBySide>
          {recipe.instructions && (
            <div>
              <RecipeLink href="#instructions" id="instructions">
                <h2>Przygotowanie</h2>
              </RecipeLink>

              <RecipeSteps>
                {recipe.instructions.map((ins, idx) => (
                  <RecipeStep
                    // eslint-disable-next-line react/no-array-index-key
                    key={idx}
                    stepNumber={idx + 1}
                    instruction={ins}
                    done={doneSteps.includes(idx)}
                    onClick={() => handleStepClicked(idx)}
                  />
                ))}
              </RecipeSteps>
            </div>
          )}

          <IngredientsSection>
            {servings && <Servings servings={servings} onServingsChange={setServings} />}

            {parsedIngredients && (
              <>
                <div>
                  <IngredientsHeader>
                    <RecipeLink href="#ingredients" id="ingredients">
                      <h2>Składniki</h2>
                    </RecipeLink>

                    <AltText>
                      {ingredientsText.format({ quantity: recipe.ingredients.length })}
                    </AltText>
                  </IngredientsHeader>

                  <IngredientsContainer>
                    {parsedIngredients.map((ingredient) => (
                      <IngredientItem ingredient={ingredient} key={ingredient.original} />
                    ))}
                  </IngredientsContainer>
                </div>

                {shoppingList && (
                  <AddToListButton icon={ShoppingList} onClick={handleAddToShoppingList}>
                    {`Dodaj do "${shoppingList.name}"`}
                  </AddToListButton>
                )}
              </>
            )}
          </IngredientsSection>
        </SideBySide>
      </ContentWrapper>
    </ThemeProvider>
  );
};
