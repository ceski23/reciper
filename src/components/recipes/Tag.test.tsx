import { renderWithProviders } from 'test/utils';

import { Tag } from 'components/recipes/Tag';

describe('Tag', () => {
  test('should render correctly', () => {
    const { asFragment } = renderWithProviders(<Tag tag="test" />);
    expect(asFragment()).toMatchSnapshot();
  });
});
