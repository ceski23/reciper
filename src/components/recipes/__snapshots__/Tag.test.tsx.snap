// Vitest Snapshot v1

exports[`Tag > should render correctly 1`] = `
<DocumentFragment>
  .emotion-1 {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #e7ecef;
  border-radius: 10px;
  padding: 10px;
  color: #0af;
  font-size: 14px;
  font-size: clamp(14px, 12.153846153846153px + 0.3205128205128205vw, 16px);
}

<a
    class="emotion-0 emotion-1 emotion-2"
    href="/tags/test"
  >
    test
  </a>
</DocumentFragment>
`;
